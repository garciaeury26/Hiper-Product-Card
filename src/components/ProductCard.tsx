import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import React, { createContext } from "react";
import { Product, ProductContextProps, onChangeArgs, InitalValues, ProductHandlers } from "../interfaces/interfaces";
import { CSSProperties } from "react"

export interface Props {
    className?: string;
    product: Product;
    // children?: ReactElement | ReactElement[]// o tambien => JSX.Element
    children: (args: ProductHandlers) => JSX.Element,
    style?: CSSProperties,
    value?: number;
    onChange?: (args: onChangeArgs) => void,
    initialValues?: InitalValues
}
export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({ children, product, value = 0, className, style, onChange, initialValues }:
    Props) => {
    const { counter, increseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    // y puedo manderle parametros
    // el children es una funcion que retorna jsx por eso la ejecuto como una funcion 

    return (
        <Provider value={{ counter, increseBy, product, isMaxCountReached }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {
                    children({
                        count: counter,
                        maxCount: maxCount,
                        isMaxCountReached,
                        product,
                        increseBy,
                        reset
                    })
                }
            </div>
        </Provider>
    );
};


