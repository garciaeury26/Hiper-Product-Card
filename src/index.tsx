
// exportamos lo que queremos servir 
export * from "./components";