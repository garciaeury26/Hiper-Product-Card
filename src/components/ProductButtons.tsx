import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";
import { CSSProperties } from "react"

export interface Props {
    className?: string;
    style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
    const { counter, increseBy, isMaxCountReached } = useContext(ProductContext);

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button onClick={() => increseBy(-1)} className={styles.buttonMinus}>
                -
            </button>
            <div className={styles.countLabel}>
                {counter}
            </div>
            <button onClick={() => increseBy(1)} className={`${!isMaxCountReached ? styles.buttonAdd : styles.disabled}`}>
                +
            </button>
        </div>
    );
};