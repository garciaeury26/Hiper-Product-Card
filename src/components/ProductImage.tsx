
import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import noImage from "../assets/no-image.jpg";
import styles from "../styles/styles.module.css";

export interface Props {
    img?: string;
    className?: string;
}

export const ProductImage = ({ img = "", className }: Props) => {
    const { product } = useContext(ProductContext);

    const src = img ? img : product.img ?? noImage;

    return (
        <img
            className={`${styles.productImg} ${className}`}
            src={src} alt="product img"
        />
    )
}
