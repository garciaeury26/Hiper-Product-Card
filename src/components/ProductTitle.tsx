import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";

export interface Props {
    title: string;
    className?: string
}

export const ProductTitle = ({ title, className }: Props) => {
    const { product } = useContext(ProductContext);

    return (
        <span className={`${styles.productDescription} ${className}`}>
            {title ? title : product.title}
        </span>
    )
}
