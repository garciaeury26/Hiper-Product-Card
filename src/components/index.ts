
import { ProductCardHOCProps } from "../interfaces/interfaces";
import { ProductButtons } from "./ProductButtons";
import { ProductCard as ProductCardHoc } from "./ProductCard";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";


// Asignar componentes a otros componentes
export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardHoc, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})

export default ProductCard;