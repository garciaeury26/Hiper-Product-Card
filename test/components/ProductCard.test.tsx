import React from 'react';
import render from "react-test-renderer"
const { act } = render;

import { ProductCard } from "../../src/components";

import { product2 } from '../data/products';

describe("Pruebas en el componente ProductImage", () => {
    test('Deve mostar el componente correctamente', () => {
        const wrapper = render.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <h1>Hola</h1>
                    )
                }
            </ProductCard>
        );
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('Deve mostrar el componente Correctamente', () => {

        const wrapper = render.create(
            <ProductCard product={product2}>
                {
                    ({ count, increseBy }) => (
                        <>
                            <h1>{count}</h1>
                            <button onClick={() => increseBy(1)}>+1</button>
                        </>
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
        console.log((tree as any).children[0].children[0]);

        // simular el increseBy
        act(() => {
            (tree as any).children[1].props.onClick();
        });

        // obtener la nuevo estado
        tree = wrapper.toJSON();
        expect((tree as any).children[0].children[0]).toBe("1");

    })
})