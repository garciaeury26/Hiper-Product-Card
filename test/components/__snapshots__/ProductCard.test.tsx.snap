// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente ProductImage Deve mostar el componente correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Hola
  </h1>
</div>
`;

exports[`Pruebas en el componente ProductImage Deve mostrar el componente Correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    0
  </h1>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;
